File: sdks/dart/lib/src/safety_number.dart
/// Sibna Protocol v3.0.1 — Safety Numbers
///
/// 80-digit human-readable fingerprints for identity verification.
/// Matches the Rust core's SafetyNumber implementation exactly:
///   - SHA-512(0x01 || "SIBNA_SAFETY_NUMBER_V1" || first_key || second_key)
///   - First 32 bytes of hash → 80 decimal digits (16 groups of 5)
///   - QR format: version(1) + "SB1"(3) + fingerprint(32) = 36 bytes
import 'dart:convert';
import 'dart:typed_data';

import 'crypto.dart';

/// Safety number for identity verification.
///
/// The safety number is derived from both parties' identity keys and
/// provides a way to detect MITM attacks during initial key exchange.
class SafetyNumber {
  /// The 80-digit string (16 groups of 5 digits, no separators).
  final String digits;

  /// The raw 32-byte fingerprint.
  final Uint8List fingerprint;

  /// Version byte.
  final int version;

  /// Current version.
  static const int currentVersion = 1;

  SafetyNumber._({
    required this.digits,
    required this.fingerprint,
    this.version = currentVersion,
  });

  /// Calculate safety number from two X25519 public keys.
  ///
  /// Keys are sorted lexicographically for consistent ordering.
  factory SafetyNumber.calculate(
    Uint8List ourIdentity,
    Uint8List theirIdentity,
  ) {
    if (ourIdentity.length != 32) {
      throw ArgumentError('Identity key must be 32 bytes');
    }
    if (theirIdentity.length != 32) {
      throw ArgumentError('Identity key must be 32 bytes');
    }

    final first = constantTimeEquals(ourIdentity, theirIdentity)
        ? ourIdentity
        : (_bytesLessThan(ourIdentity, theirIdentity)
            ? ourIdentity
            : theirIdentity);
    final second = constantTimeEquals(ourIdentity, theirIdentity)
        ? theirIdentity
        : (_bytesLessThan(ourIdentity, theirIdentity)
            ? theirIdentity
            : ourIdentity);

    final data = Uint8List.fromList([
      currentVersion,
      ...utf8.encode('SIBNA_SAFETY_NUMBER_V1'),
      ...first,
      ...second,
    ]);

    final hash = sha512Hash(data);
    final fp = Uint8List.fromList(hash.sublist(0, 32));
    final d = _bytesToDigits(fp);

    return SafetyNumber._(
      digits: d,
      fingerprint: fp,
    );
  }

  /// Calculate safety number with additional data (e.g. for group verification).
  factory SafetyNumber.calculateWithExtra(
    Uint8List ourIdentity,
    Uint8List theirIdentity,
    Uint8List extraData,
  ) {
    if (ourIdentity.length != 32) {
      throw ArgumentError('Identity key must be 32 bytes');
    }
    if (theirIdentity.length != 32) {
      throw ArgumentError('Identity key must be 32 bytes');
    }

    final first = _bytesLessThan(ourIdentity, theirIdentity)
        ? ourIdentity
        : theirIdentity;
    final second = _bytesLessThan(ourIdentity, theirIdentity)
        ? theirIdentity
        : ourIdentity;

    final data = Uint8List.fromList([
      currentVersion,
      ...utf8.encode('SIBNA_SAFETY_NUMBER_V1_EXTRA'),
      ...first,
      ...second,
      ...extraData,
    ]);

    final hash = sha512Hash(data);
    final fp = Uint8List.fromList(hash.sublist(0, 32));
    final d = _bytesToDigits(fp);

    return SafetyNumber._(
      digits: d,
      fingerprint: fp,
    );
  }

  /// Parse safety number from its string representation.
  factory SafetyNumber.fromString(String s) {
    final digitsStr = s.replaceAll(RegExp(r'\D'), '');
    if (digitsStr.length != 80) {
      throw ArgumentError('Expected 80 digits, got ${digitsStr.length}');
    }
    final fp = _digitsToBytes(digitsStr);
    return SafetyNumber._(
      digits: digitsStr,
      fingerprint: fp,
    );
  }

  /// Return the 80-digit string (no separators).
  String asString() => digits;

  /// Return the safety number formatted for display.
  ///
  /// [groupSize]: digits per group (default 5)
  /// [groupsPerLine]: groups per line (default 8)
  String formatted({int groupSize = 5, int groupsPerLine = 8}) {
    final groups = <String>[];
    for (int i = 0; i < digits.length; i += groupSize) {
      final end = (i + groupSize < digits.length) ? i + groupSize : digits.length;
      groups.add(digits.substring(i, end));
    }

    final lines = <String>[];
    for (int i = 0; i < groups.length; i += groupsPerLine) {
      final end = (i + groupsPerLine < groups.length)
          ? i + groupsPerLine
          : groups.length;
      lines.add(groups.sublist(i, end).join(' '));
    }
    return lines.join('\n');
  }

  /// QR code data: version(1) + "SB1"(3) + fingerprint(32) = 36 bytes.
  Uint8List qrData() {
    return Uint8List.fromList([
      version,
      ...utf8.encode('SB1'),
      ...fingerprint,
    ]);
  }

  /// Constant-time comparison with another safety number.
  bool verify(SafetyNumber other) {
    return constantTimeEquals(fingerprint, other.fingerprint);
  }

  @override
  String toString() => formatted();

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    if (other is! SafetyNumber) return false;
    return verify(other);
  }

  @override
  int get hashCode => fingerprint.hashCode;

  // ── Internal helpers ────────────────────────────────────────────────

  /// Convert 32 bytes to 80 decimal digits (16 chunks × 5 digits).
  static String _bytesToDigits(Uint8List data) {
    final parts = <String>[];
    for (int i = 0; i < 32; i += 2) {
      final value = (data[i] << 8) | data[i + 1];
      parts.add('${value % 100000}'.padLeft(5, '0'));
    }
    return parts.join('');
  }

  /// Convert 80 digits back to 32 bytes.
  static Uint8List _digitsToBytes(String digits) {
    final result = Uint8List(32);
    var byteIdx = 0;
    for (int i = 0; i < 80; i += 5) {
      final chunk = digits.substring(i, i + 5);
      final value = int.parse(chunk);
      result[byteIdx] = (value >> 8) & 0xFF;
      result[byteIdx + 1] = value & 0xFF;
      byteIdx += 2;
    }
    return result;
  }

  /// Lexicographic byte comparison.
  static bool _bytesLessThan(Uint8List a, Uint8List b) {
    final minLen = a.length < b.length ? a.length : b.length;
    for (int i = 0; i < minLen; i++) {
      if (a[i] < b[i]) return true;
      if (a[i] > b[i]) return false;
    }
    return a.length < b.length;
  }
}

File: sdks/javascript/src/safety_number.ts
/**
 * Sibna Protocol v3.0.1 — Safety Numbers (JavaScript)
 */

import { sha512Hash, hmacSha256 } from './crypto';

export class SafetyNumber {
  readonly digits: string;
  readonly fingerprint: Uint8Array;
  readonly version: number;

  static readonly VERSION = 1;

  private constructor(digits: string, fingerprint: Uint8Array, version = 1) {
    this.digits = digits;
    this.fingerprint = fingerprint;
    this.version = version;
  }

  static calculate(ourIdentity: Uint8Array, theirIdentity: Uint8Array): SafetyNumber {
    const [first, second] = compareBytes(ourIdentity, theirIdentity)
      ? [ourIdentity, theirIdentity]
      : [theirIdentity, ourIdentity];

    const hasher = new Uint8Array([
      SafetyNumber.VERSION,
      ...new TextEncoder().encode('SIBNA_SAFETY_NUMBER_V1'),
      ...first,
      ...second,
    ]);
    const hash = sha512Hash(hasher);
    const fingerprint = hash.slice(0, 32);
    const digits = bytesToDigits(fingerprint);

    return new SafetyNumber(digits, fingerprint);
  }

  static calculateWithExtra(
    ourIdentity: Uint8Array, theirIdentity: Uint8Array, extraData: Uint8Array,
  ): SafetyNumber {
    const [first, second] = compareBytes(ourIdentity, theirIdentity)
      ? [ourIdentity, theirIdentity]
      : [theirIdentity, ourIdentity];

    const parts = new Uint8Array([
      SafetyNumber.VERSION,
      ...new TextEncoder().encode('SIBNA_SAFETY_NUMBER_V1_EXTRA'),
      ...first, ...second, ...extraData,
    ]);
    const hash = sha512Hash(parts);
    const fingerprint = hash.slice(0, 32);
    return new SafetyNumber(bytesToDigits(fingerprint), fingerprint);
  }

  static fromString(s: string): SafetyNumber {
    const digits = s.replace(/\D/g, '');
    if (digits.length !== 80) throw new Error(`Expected 80 digits, got ${digits.length}`);
    const fingerprint = digitsToBytes(digits);
    return new SafetyNumber(digits, fingerprint);
  }

  asString(): string { return this.digits; }

  formatted(groupSize = 5, groupsPerLine = 8): string {
    const groups: string[] = [];
    for (let i = 0; i < this.digits.length; i += groupSize) {
      groups.push(this.digits.substring(i, i + groupSize));
    }
    const lines: string[] = [];
    for (let i = 0; i < groups.length; i += groupsPerLine) {
      lines.push(groups.slice(i, i + groupsPerLine).join(' '));
    }
    return lines.join('\n');
  }

  qrData(): Uint8Array {
    const qr = new Uint8Array(36);
    qr[0] = this.version;
    qr.set(new TextEncoder().encode('SB1'), 1);
    qr.set(this.fingerprint, 4);
    return qr;
  }

  verify(other: SafetyNumber): boolean {
    const k = new Uint8Array(32);
    const a = hmacSha256(k, this.fingerprint);
    const b = hmacSha256(k, other.fingerprint);
    return a.every((v, i) => v === b[i]);
  }

  equals(other: SafetyNumber): boolean {
    if (this.fingerprint.length !== other.fingerprint.length) return false;
    let diff = 0;
    for (let i = 0; i < this.fingerprint.length; i++) {
      diff |= this.fingerprint[i] ^ other.fingerprint[i];
    }
    return diff === 0;
  }
}

function compareBytes(a: Uint8Array, b: Uint8Array): boolean {
  for (let i = 0; i < a.length; i++) {
    if (a[i] < b[i]) return true;
    if (a[i] > b[i]) return false;
  }
  return false;
}

function bytesToDigits(bytes: Uint8Array): string {
  const parts: string[] = [];
  for (let i = 0; i < 32; i += 2) {
    const value = (bytes[i] << 8) | bytes[i + 1];
    parts.push(String(value % 100000).padStart(5, '0'));
  }
  return parts.join('');
}

function digitsToBytes(digits: string): Uint8Array {
  const result = new Uint8Array(32);
  for (let i = 0; i < 80; i += 5) {
    const value = parseInt(digits.substring(i, i + 5));
    const byteIdx = (i / 5) * 2;
    result[byteIdx] = (value >> 8) & 0xFF;
    result[byteIdx + 1] = value & 0xFF;
  }
  return result;
}
